From 54f6f59117a7113fa9da18d4fb1ca275faa508a9 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Fri, 23 Jan 2015 18:35:35 +0800
Subject: [PATCH 0077/5242] MLK-10132-3 usb: chipidea: udc: do not enter low
 power mode if vbus on

commit  c1deea26cbac47be1c278fb1bf49db7f84a7410d from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch is to prevent usb entering low power mode if vbus is on even gadget
driver is not binded, by holding the PM count of ci->dev.
So, there are 3 pm usage_count status:
- ci->dev: 1 ci->gadget.dev: 1
  Device mode with gadget driver binded and vbus on.
- ci->dev: 1 ci->gadget.dev: 0
  USB vbus on but gadget driver not binded.
- ci->dev: 0 ci->gadget.dev: 1
  USB OTG FSM is in a_peripheral mode.
Above 2 device's pm usage_count hold by ci otg(ci->dev) and usb gadget
(ci->gadget.dev).

Signed-off-by: Li Jun <jun.li@freescale.com>
(cherry picked from commit 673c6bf1b3aa0b1b698569b9259712b0e765be32)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/udc.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index 79930ce..5073d8d 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -1955,8 +1955,10 @@ int ci_usb_charger_connect(struct ci_hdrc *ci, int is_active)
 {
 	int ret = 0;
 
+	if (is_active)
+		pm_runtime_get_sync(ci->dev);
+
 	if (ci->platdata->notify_event) {
-		pm_runtime_get_sync(&ci->gadget.dev);
 		if (is_active)
 			hw_write(ci, OP_USBCMD, USBCMD_RS, 0);
 
@@ -1971,8 +1973,11 @@ int ci_usb_charger_connect(struct ci_hdrc *ci, int is_active)
 			/* Pull down dp */
 			hw_write(ci, OP_USBCMD, USBCMD_RS, 0);
 		}
-		pm_runtime_put_sync(&ci->gadget.dev);
 	}
+
+	if (!is_active)
+		pm_runtime_put_sync(ci->dev);
+
 	return ret;
 }
 
-- 
1.7.9.5

